File: 1558-course-schedule-iv/1558-course-schedule-iv.java
class Solution {
    public List<Boolean> checkIfPrerequisite(int numCourses, int[][] prerequisites, int[][] queries) {
        int n = numCourses;
        int p = prerequisites.length;
        boolean[][] matrix = new boolean[n][n];

        for (int[] prerequisite : prerequisites) {
            int from = prerequisite[0];
            int to = prerequisite[1];

            matrix[from][to] = true;
        }

        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    matrix[i][j] = matrix[i][j] || matrix[i][k] && matrix[k][j];
                }
            }
        }

        List<Boolean> ans = new ArrayList<>();

        for (int[] query : queries) {
            int from = query[0];
            int to = query[1];

            ans.add(matrix[from][to]);
        }

        return ans;
    }
}

File: 1558-course-schedule-iv/1558-course-schedule-iv.py
class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:

        preMap = defaultdict(set)
        indegrees = [0] * numCourses
        adj = [[] for _ in range(numCourses)]

        for pre, next in prerequisites:
            adj[pre].append(next)
            indegrees[next] += 1

        q = deque([i for i in range(numCourses) if indegrees[i] == 0])

        while q:
            course = q.popleft()
            for next in adj[course]:
                indegrees[next] -= 1
                preMap[next].add(course)
                preMap[next].update(preMap[course])
                if indegrees[next] == 0:
                    q.append(next)
                    
        return [a in preMap[b] for a, b in queries]

                